### PR TITLE
Fix 'Add to calendar' link for Outlook.com

### LINKED ISSUE #83: Microsoft does not like the normal space character anymore :shrug:
Quite an odd issue... I noticed that strings with spaces (event title or description) result in weird characters in Outlook. For instance, space is rendered as `+`. Even if the text is encoded, and spaces are replaced with `%20` after encoding.

Demo: https://outlook.live.com/calendar/0/deeplink/compose?body=DESCRI%C3%87%C3%83O%20MUITO%20GRANDE&enddt=2021-03-17T20%3A45%3A00%2B00%3A00&location=&path=%2Fcalendar%2Faction%2Fcompose&rru=addevent&startdt=2021-03-17T17%3A15%3A00%2B00%3A00&subject=TITULO%20MUITO%20GRANDE

However, replacing the normal spaces with a thin space solves the problem! :exploding_head: The link below is the same as the first one. However, normal spaces (`%20`) are replaced with thin spaces (`%E2%80%89`): https://outlook.live.com/calendar/0/deeplink/compose?body=DESCRI%C3%87%C3%83O%E2%80%89MUITO%E2%80%89GRANDE&enddt=2021-03-17T20%3A45%3A00%2B00%3A00&location=&path=%2Fcalendar%2Faction%2Fcompose&rru=addevent&startdt=2021-03-17T17%3A15%3A00%2B00%3A00&subject=TITULO%E2%80%89MUITO%E2%80%89GRANDE

I am still trying to figure out why... and if other characters are affected.

### LINKED ISSUE #84: 'Add to Calendar' links for Outlook is out of date
New links for Outlook.com and Office 365 are available

### LINKED ISSUE #85: URI encoding not working in `Extractor`
Example from `calendar.js`.

```js
console.log(extractor.getName(event, true));
console.log(encodeURIComponent(extractor.getName(event, true)));
```

Output:
```
[AMAT2] - T - EaD
%5BAMAT2%5D%20-%20T%20-%20EaD
```

Notice the second parameter for `extractor.getName` is `true`, hence the return should be encoded.

### PR DESCRIPTION
**Pull request by fabiodrg**
- [x] Update the link for Outlook.com events #84 
- [x] Add support for Office365 events
- [x] Refactor the functions in `calendar.js` to make it more maintainable and document the module with JSDoc (I am thinking of hosting a documentation website in GitHub pages. We can use GitHub actions and the `jsdoc` tool to automate the generation. I did some experiments, seems good enough. Just food for thought)
- [x] Handle the weird Microsoft encoding problem #83 
- [x] Fix the URI encoding bug in `Extractor` #85 

Closes #83, closes #84, closes #85

### CODE REVIEW

File: src/js/utils/calendar.js [hunk 1]
```diff
@@ -1,38 +1,91 @@
...
+ *  * Standard: {@link https://icalendar.org/iCalendar-RFC-5545/3-8-5-3-recurrence-rule.html}
+ *  * Date is in the same format as start/end dates
+
  * @param {Extractor} extractor class that implements getName and getDescription from the event
  * @param {event object} event needs to have (at least) {from, to, location, download}
- * @param {*} repeat if undefined the even does not repeat overtime, otherwise it does (uses the same format as ics.js, so: repeat = { freq: "WEEKLY", until: stringFriendlyWithDate };)
- * https://richmarr.wordpress.com/2008/01/07/adding-events-to-users-calendars-part-2-web-calendars/
+ * @param {*} repeat
  */
 function eventToGCalendar(extractor, event, repeat) {
```
[the remark this answers is not in the document]
Comment (fabiodrg) [reply src/js/utils/calendar.js:38]: Fixed

File: src/js/utils/calendar.js [hunk 2]
```diff
@@ -1,105 +1,176 @@
...
-    let dates = '';
-    if(event.from && event.to)
-        dates = `&dates=${event.from.toGCalendar()}/${event.to.toGCalendar()}`;
-    
-    if(event)
-    return (`https://calendar.google.com/calendar/r/eventedit?text=${extractor.getName(event, true)}&location=${event.location}&details=${extractor.getDescription(event, true, false)}&sprop=name:${extractor.getName(event, true)}&sprop=website:${"https://github.com/msramalho/SigTools"}${recur}${dates}`);
+    return {
+        rru: 'addevent',
+        path: '/calendar/action/compose',
+        subject: encodeURIComponent(customEncode(extractor.getName(event, true))),
```
Comment (fabiodrg) [inline src/js/utils/calendar.js:95]: This `encodeURIComponent` is due to #85 . Should be removed once the issue is fixed

Comment (fabiodrg) [reply src/js/utils/calendar.js:95]: Yahoo and Google don't seem to require encoding as far as I can tell, hence I only added it here

File: src/js/utils/modal.js [hunk 3]
```diff
@@ -84,7 +84,8 @@ function getRepeat(from, to) {
     if (from && to && dayDiff(from, to) > 6) {
         repeat = {
             freq: "WEEKLY",
-            until: to.toString()
+            //until: to.toString()
```
[the remark this answers is not in the document]
Comment (fabiodrg) [reply src/js/utils/modal.js:87]: As far as I can tell, it is only used in `calendar.js`, so it's safe

**Comment by fabiodrg** [comment]
Hi @msramalho ! Do you have a few minutes to review this MR?

Just one note. In order to fix #85, I am no longer using the `convertToURI` defined in the `Extractor` class. I noticed that calling either `getDescription()` or `getName()` did not return a fully encoded URL, even if the parameter `forUrl` was `true`. The reason is `convertToURI()` works per object field. However, when the user-defined template is evaluated with `parseStrFormat()`, the final string is not encoded. Therefore, if my description template is `${param1} something bla bla %? ${param2}`, both parameters are encoded, but the spaces and other special characters are not. At this point, I wondered if we could simplify the flow: call `parseStrFormat()` and then encode the final string. Do you think there's any problem with this? I really can't remember the reasoning for the other approach :confused:  If this approach is fine, then some functions in `Extractor` can be deprecated. Also, to I made the interface more friendly with `getName` and a encoded variant `getNameEncoded`. I think makes the code more readable than passing boolean parameters. Let me know if you agree :smile:

File: src/js/utils/calendar.js [hunk 4]
```diff
@@ -1,105 +1,213 @@
...
+     * * Unicode for thin space: `U+2009`
+     * 
+     * Note: replaceAll() for strings still not well supported, and using
+     * unicode in regex (to replace all ocurrences with replace()) is tricky
+     * as well. Hacky solution below uses split() and join()
+     * 
+     * @see {@link https://github.com/msramalho/SigTools/issues/83}
+     * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll}
+     */
+    const customEncode = (str) => str.split('\u0020').join('\u2009');
```
Comment (msramalho) [inline src/js/utils/calendar.js:124]: For some reason, this conversion is being interpreted as a `+` in outlook: 
image

do you have any guesses as to why that might be?

**Review by msramalho — APPROVED** [review]
@fabiodrg Great work, and I see a lot of effort put into proper documentation as well 👏 🙇‍♂️ 

I made some minor updates, bumped the version with `npm version minor -m "bump version"` and some minor code changes too.

**Pull request merged**